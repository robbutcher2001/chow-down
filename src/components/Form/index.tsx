import React, { Component, FormEvent, ChangeEvent, ReactElement } from 'react';

import styled from 'styled-components';

import { InputBoxProps } from '../InputBox';

interface FieldNames {
  [key: string]: string
};

interface StateProps { };

interface DispatchProps { };

interface OwnProps {
  dispatch: (form: object) => object,
  submitText: string,
  children: ReactElement<InputBoxProps> | ReactElement<InputBoxProps>[]
};

interface OwnState {
  form: {
    [key: string]: string | number
  }
};

type CombinedProps = StateProps & DispatchProps & OwnProps;

const Form = styled.form`
  display: flex;
  flex-direction: column;

  > .button-group {
    display: grid;
    grid-template-columns: repeat(auto-fit, minmax(280px, 1fr));
    gap: 0 2em;

    > button {
      border-radius: 5px;
      padding: 0 0.5em;
      margin: 0.5rem 0;
      height: 3rem;
      font-size: 1rem;
      border: none;
      color: white;
      cursor: pointer;
  
      &[type=submit] {
        background-color: #4acaa8;
      }
  
      &[type=reset] {
        background-color: #989898;
      }
    }
  }
`

class FormComponent extends Component<CombinedProps, OwnState> {
  constructor(props: CombinedProps) {
    super(props);
    const fieldNames: FieldNames = this.getFieldNames();

    this.state = {
      form: {
        ...fieldNames
      }
    };
  }

  getFieldNames = () => React.Children.map(this.props.children, (child: ReactElement<InputBoxProps>) => child.props.name)
    .reduce((names, name) => {
      names[name] = '';
      return names;
    }, {} as FieldNames);

  onChange = (
    field: string,
    type: 'text' | 'number',
    event: ChangeEvent<HTMLInputElement> | ChangeEvent<HTMLTextAreaElement>
  ) => {
    event.preventDefault();
    const value = event.currentTarget.value;
    this.setState(prevState => {
      const newState = {
        form: Object.assign({}, prevState.form)
      };
      newState.form[field] = type === 'number' && value ? parseInt(value) : value;
      return newState;
    });
  }

  onSubmit = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    const formPopulated = Object.keys(this.state.form).reduce((acc, formField) =>
      this.state.form[formField] ? true : acc, false);

    if (formPopulated) {
      this.props.dispatch(this.state.form);
    }
  }

  onReset = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    const fieldNames: FieldNames = this.getFieldNames();

    this.setState({
      form: {
        ...fieldNames
      }
    });
  };

  render = () => {
    const children = React.Children.map(this.props.children, (child: ReactElement<InputBoxProps>) =>
      React.cloneElement(child, { form: this.state.form, onChange: this.onChange }));

    return (
      <Form id='form' onSubmit={this.onSubmit} onReset={this.onReset} >
        {children}
        <div className='button-group'>
          <button type='submit' form='form' value={this.props.submitText}>
            {this.props.submitText}
          </button>
          <button type='reset' form='form' value='Reset'>
            Reset
          </button>
        </div>
      </Form>
    );
  }
};

export default FormComponent;