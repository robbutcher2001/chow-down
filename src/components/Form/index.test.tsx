import React from 'react';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';

import Form from '.';

test('Form basic snapshot render', () => {
  const form = renderer.create(
    <Form name='fakeForm' dispatch={jest.fn()} submitText='Submit Button' >
      <div>fake_child</div>
    </Form>
  );

  expect(form.toJSON()).toMatchSnapshot();
});

test('Form correct tag content assertion', () => {
  const { getByText } = render(
    <Form name='fakeForm' dispatch={jest.fn()} submitText='Submit Button' >
      <div>fake_child</div>
    </Form>
  );

  expect(getByText(/Submit Button/).textContent).toEqual('Submit Button');
  expect(getByText(/Reset/).textContent).toEqual('Reset');
});