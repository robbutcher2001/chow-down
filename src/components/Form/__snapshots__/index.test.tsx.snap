// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form basic snapshot render 1`] = `
<form
  className="sc-bdVaJa bgwYpC"
  id="form"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    form={Object {}}
    setNewFormState={[Function]}
  >
    fake_child
  </div>
  <div
    className="button-group"
  >
    <button
      form="form"
      type="submit"
      value="Submit Button"
    >
      Submit Button
    </button>
    <button
      form="form"
      type="reset"
      value="Reset"
    >
      Reset
    </button>
  </div>
</form>
`;
